### PR TITLE
quic: moving helpers to be shared by connection pool

### PR DESCRIPTION
**Pull request by alyssawilk**
This fixes a TODO and also the tsan issues triggered with the last merge.

Risk Level: n/a (quic upstream only)
Testing: integration tests
Docs Changes: n/a
Release Notes: n/a

### CODE REVIEW

File: source/common/http/http3/quic_client_connection_factory.h [hunk 1]
```diff
@@ -10,17 +10,28 @@
...
+
 // A factory to create EnvoyQuicClientSession and EnvoyQuicClientConnection for QUIC
 class QuicClientConnectionFactory : public Config::UntypedFactory {
 public:
   ~QuicClientConnectionFactory() override = default;
 
-  virtual std::unique_ptr<Network::ClientConnection>
-  createQuicNetworkConnection(Network::Address::InstanceConstSharedPtr server_addr,
-                              Network::Address::InstanceConstSharedPtr local_addr,
+  // Allows thread-local creation of PersistentQuicInfo.
```
Comment (danzh2010) [inline source/common/http/http3/quic_client_connection_factory.h:24]: I can understand this object is per-conn pool. But is it thread-local? 
Plz add comment how to use these interfaces to create a client connection.

**Review by danzh2010 — DISMISSED** [review]
Thanks for fixing the test! LGTM other than a request for mroe documentation.

File: source/common/http/http3/quic_client_connection_factory.h (hunk 1 above)
Comment (alyssawilk) [reply source/common/http/http3/quic_client_connection_factory.h:24]: connection pools must be thread local (they're not thread safe).  will comment :-)

File: source/common/http/http3/conn_pool.cc [hunk 2]
```diff
@@ -17,30 +17,62 @@ namespace Envoy {
...
+                              random_generator, state, client_fn, codec_fn, protocol) {
+    auto source_address = host_->cluster().sourceAddress();
+    if (!source_address.get()) {
+      auto host_address = host->address();
+      source_address = Network::Utility::getLocalAddress(host_address->ip()->version());
+    }
+    Network::TransportSocketFactory& transport_socket_factory = host->transportSocketFactory();
+    quic_info_ =
+        Config::Utility::getAndCheckFactoryByName<Http::QuicClientConnectionFactory>(
+            Http::QuicCodecNames::get().Quiche)
```
Comment (mattklein123) [inline source/common/http/http3/conn_pool.cc:41]: example of cross dependency issue that we were discussing in the slack thread (so yeah, I think I'm firmly in the no extension camp now!)

**Review by mattklein123 — DISMISSED** [review]
LGTM!

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**